src/entities/swap/swaps/v2/auraBalSwaps/joinSwap.ts

```diff
@@ -0,0 +1,133 @@
+import { Relayer } from '@/entities/relayer';
+import { auraBalToken, balWethAddress, auraBAL } from './constants';
+import { getJoinData } from './joinPool';
+import { getSwapData } from './swap';
+import { AuraBalSwapQueryOutput } from './auraBalSwaps';
+import { Token } from '@/entities/token';
+import { AuraBalSwapQueryInput } from './parseInputs';
+
+// token[join]8020BPT[swap]auraBAL
+export async function queryJoinSwap(
+    input: AuraBalSwapQueryInput & { client },
+): Promise<AuraBalSwapQueryOutput> {
+    const { swapToken: joinToken, inputAmount, kind, client } = input;
+    const value = 0n;
+
+    // join BAL-WETH 80/20 Pool with joinToken and get 8020BPT in return
+    const { joinPoolData, joinPoolOpRef } = getJoinData(
+        joinToken,
+        zeroAddress, // Note zeroAddress used for query but not for build
+        inputAmount.amount,
+        false, // for query we always use WETH
+    );
+
+    // swap 8020BPT>aurABL through auraBal/8020BPT stable pool
+    // For query we set limit to 0
+    const { swapData, swapOpRef } = getSwapData(
+        joinPoolOpRef,
+        balWethAddress,
+        auraBAL as Address,
+        zeroAddress,
+        zeroAddress,
+        0n,
+        value,
+        true,
+    );
+
+    // peek call is used to read final result
+    const peekData = Relayer.encodePeekChainedReferenceValue(swapOpRef);
+
+    // vaultActionsQueryMulticall allows us to query even if user has no balance/allowance
+    const encodedMulticall = encodeFunctionData({
+        abi: balancerRelayerAbi,
+        functionName: 'vaultActionsQueryMulticall',
+        args: [[joinPoolData, swapData, peekData]],
+    });
+
+    const { data } = await client.call({
+        to: BALANCER_RELAYER[1],
+        data: encodedMulticall,
+    });
+
+    const result = decodeFunctionResult({
+        abi: balancerRelayerAbi,
+        functionName: 'vaultActionsQueryMulticall',
+        data: data as Hex,
+    });
+
+    const peekedValue = decodeAbiParameters(
+        [{ type: 'uint256' }],
+        result[result.length - 1],
+    )[0];
+    return {
+        inputAmount: TokenAmount.fromRawAmount(joinToken, inputAmount.amount),
+        expectedAmountOut: TokenAmount.fromRawAmount(auraBalToken, peekedValue),
+        kind,
+    };
+}
+
+export function buildJoinSwapCall(
+    userAddress: Address,
+    inputAmount: bigint,
+    swapLimit: bigint,
+    joinToken: Token,
+    wethIsEth: boolean,
+    relayerApprovalSignature?: Hex,
+): { callData: Hex; value: bigint } {
+    // join BAL-WETH 80/20 Pool with joinToken and get 8020BPT in return (to the RELAYER)
+    const { joinPoolData, joinPoolOpRef, value } = getJoinData(
+        joinToken,
+        userAddress,
+        inputAmount,
+        wethIsEth,
+    );
+
+    // Older pools don't have pre-approval so need to add this as a step, approves Vault to spend on RELAYERS behalf
+    const approval = Relayer.encodeApproveVault(balWethAddress, joinPoolOpRef);
```

Comment:
This one took me a while 😅 - I had to go to the old code to understand it.
I didn't know about it and I never realized this was present in the old code 👀 
Not sure how useful that info is from now on, but now I know 😂